Add plugin: Theme Engine

YazanAmmar opened the pull request:
# I am submitting a new Community Plugin

- [x] I attest that I have done my best to deliver a high-quality plugin, am proud of the code I have written, and would recommend it to others. I commit to maintaining the plugin and being responsive to bug reports. If I am no longer able to maintain it, I will make reasonable efforts to find a successor maintainer or withdraw the plugin from the directory.

## Repo URL

Link to my plugin: https://github.com/YazanAmmar/obsidian-theme-engine

## Release Checklist
- [x] I have tested the plugin on
  - [x]  Windows
  - [x]  macOS
  - [x]  Linux
  - [x]  Android _(if applicable)_
  - [x]  iOS _(if applicable)_
- [x] My GitHub release contains all required files (as individual files, not just in the source.zip / source.tar.gz)
  - [x] `main.js`
  - [x] `manifest.json`
  - [x] `styles.css`
- [x] GitHub release name matches the exact version number specified in my manifest.json (`2.0.0`)
- [x] The `id` in my `manifest.json` matches the `id` in the `community-plugins.json` file.
- [x] My README.md describes the plugin's purpose and provides clear usage instructions.
- [x] I have read the developer policies at https://docs.obsidian.md/Developer+policies, and have assessed my plugins's adherence to these policies.
- [x] I have read the tips in https://docs.obsidian.md/Plugins/Releasing/Plugin+guidelines and have self-reviewed my plugin to avoid these common pitfalls.
- [x] I have added a license in the LICENSE file.
- [x] My project respects and is compatible with the original license of any code from other plugins that I'm using. I have given proper attribution to these other projects in my `README.md`.

YazanAmmar commented:
/skip

Hi there! I'm skipping the automated check for the "avoid assigning styles via JavaScript" rule for the following reason:

The core functionality of the "Color Master" plugin is to allow users to dynamically change Obsidian's UI colors by manipulating CSS variables (custom properties) at runtime.

The detected uses of `element.style.setProperty` and `element.style.removeProperty` are the correct and modern standard for dynamically setting and unsetting CSS variables on the document body or specific elements. This is essential for the plugin to work as intended and is the same pattern used by other popular theming plugins like Style Settings.

Replacing this dynamic functionality with static CSS classes would defeat the purpose of the plugin. The styles being applied are CSS variables (e.g., `--background-primary`), which are designed to be manipulated via JavaScript, making the plugin fully compatible and adaptable with user themes.

"The remaining document.body.style calls are the core function of the plugin, which is to dynamically set and remove global CSS variables for live theming. The other .style calls are necessary for overriding another plugin's (Iconize) inline styles."

YazanAmmar commented:
> > The detected uses of element.style.setProperty and element.style.removeProperty are the correct and modern standard for dynamically setting and unsetting CSS variables on the document body or specific elements. This is essential for the plugin to work as intended and is the same pattern used by other popular theming plugins like Style Settings.
> 
> You should use `.setCssProps` instead, for example, `el.setCssProps({ "--my-width", "20px" });`. This is the recommended approach.
> 
> I'm going to reassign the bot to this, once you've made those adjustments it should get back in the human reviewer queue.

Done

Zachatoo commented:
[# Obsidian Theme Engine](https://github.com/YazanAmmar/obsidian-theme-engine/blob/d43321f7c122433735dea80e43ee56e7680334a1/README.md#L1)
Naming something "Obsidian xyz" is reserved for first party products we create, please change this header.

[Copyright (c) 2025 Yazan Ammar](https://github.com/YazanAmmar/obsidian-theme-engine/blob/d43321f7c122433735dea80e43ee56e7680334a1/LICENSE#L3)
Please update this to the current year

['obsidian-theme-engine',](https://github.com/YazanAmmar/obsidian-theme-engine/blob/d43321f7c122433735dea80e43ee56e7680334a1/src/core/migrations.ts#L7-L9)
Please don't mess with users hotkeys or community-plugins.json on plugin load. If you have current users that need to update then you'll need to ask them to setup relevant hotkeys again.
I'd recommend deleting this whole file. After your plugin is accepted and you need to make breaking changes to your plugin data schema, you can add migration logic for that (which you already have elsewhere), but please don't add migration logic for switching between plugin ids.

[const langCode = settings.language as LocaleCode;](https://github.com/YazanAmmar/obsidian-theme-engine/blob/d43321f7c122433735dea80e43ee56e7680334a1/src/i18n/strings.ts#L35), [const obsidianLang = moment.locale();](https://github.com/YazanAmmar/obsidian-theme-engine/blob/d43321f7c122433735dea80e43ee56e7680334a1/src/main.ts#L335), ["minAppVersion": "1.5.0",](https://github.com/YazanAmmar/obsidian-theme-engine/blob/d43321f7c122433735dea80e43ee56e7680334a1/manifest.json#L5), [const languageSetting = createTopSetting()](https://github.com/YazanAmmar/obsidian-theme-engine/blob/d43321f7c122433735dea80e43ee56e7680334a1/src/ui/settings/render.ts#L90-L92)
This should not be a setting, use the `getLanguage()` function from the API to get the language configured in the Obsidian settings (requires `minAppVersion: "1.8.0"`)

[heading: 'Advanced settings',](https://github.com/YazanAmmar/obsidian-theme-engine/blob/d43321f7c122433735dea80e43ee56e7680334a1/src/i18n/locales/en.ts#L65), [backgroundModalSettingsTitle: 'Background settings',](https://github.com/YazanAmmar/obsidian-theme-engine/blob/d43321f7c122433735dea80e43ee56e7680334a1/src/i18n/locales/en.ts#L87)
[Avoid "settings" in settings headings](https://docs.obsidian.md/Plugins/Releasing/Plugin+guidelines#Avoid+%22settings%22+in+settings+headings)

[const backgroundsPath = `${plugin.app.vault.configDir}/backgrounds`;](https://github.com/YazanAmmar/obsidian-theme-engine/blob/d43321f7c122433735dea80e43ee56e7680334a1/src/core/background-media.ts#L154)
Please store this data in your plugin's config folder instead, using [`<plugin instance>.manifest.dir`](https://docs.obsidian.md/Reference/TypeScript+API/PluginManifest/dir) instead of `<vault instance>.configDir`

[callback: async () => {](https://github.com/YazanAmmar/obsidian-theme-engine/blob/d43321f7c122433735dea80e43ee56e7680334a1/src/commands/index.ts#L42), [callback: async () => {](https://github.com/YazanAmmar/obsidian-theme-engine/blob/d43321f7c122433735dea80e43ee56e7680334a1/src/commands/index.ts#L60), [callback: async () => {](https://github.com/YazanAmmar/obsidian-theme-engine/blob/d43321f7c122433735dea80e43ee56e7680334a1/src/commands/index.ts#L91)
This should be a [conditional command](https://docs.obsidian.md/Plugins/User+interface/Commands#Conditional+commands)

[export function debounce(fn: (...args: unknown[]) => void, ms = 200) {](https://github.com/YazanAmmar/obsidian-theme-engine/blob/d43321f7c122433735dea80e43ee56e7680334a1/src/utils.ts#L75-L81), [debounce(func: (...args: unknown[]) => void, wait: number) {](https://github.com/YazanAmmar/obsidian-theme-engine/blob/d43321f7c122433735dea80e43ee56e7680334a1/src/ui/modals/settings/background.ts#L171-L183), [\_debounce(func: (...args: unknown[]) => void, delay: number) {](https://github.com/YazanAmmar/obsidian-theme-engine/blob/d43321f7c122433735dea80e43ee56e7680334a1/src/ui/modals/editors/snippet-css.ts#L28-L34), and others
Use [`debounce`](https://docs.obsidian.md/Reference/TypeScript+API/debounce) from the Obsidian API instead.

[this.colorUpdateInterval = window.setInterval(() => {](https://github.com/YazanAmmar/obsidian-theme-engine/blob/d43321f7c122433735dea80e43ee56e7680334a1/src/main.ts#L95)
Events need to be registered, so they can be properly unloaded when the plugin is disabled. <https://docs.obsidian.md/Plugins/Events>

[.is-ltr {](https://github.com/YazanAmmar/obsidian-theme-engine/blob/d43321f7c122433735dea80e43ee56e7680334a1/src/styles/base/_direction.scss#L2), [.is-rtl {](https://github.com/YazanAmmar/obsidian-theme-engine/blob/d43321f7c122433735dea80e43ee56e7680334a1/src/styles/base/_direction.scss#L22)
Please update your classes to be more specific to your plugin to avoid conflicts with other plugins.

[.tag {](https://github.com/YazanAmmar/obsidian-theme-engine/blob/d43321f7c122433735dea80e43ee56e7680334a1/src/styles/base/_editor.scss#L51), [.cm-s-obsidian span.cm-hashtag {](https://github.com/YazanAmmar/obsidian-theme-engine/blob/d43321f7c122433735dea80e43ee56e7680334a1/src/styles/base/_editor.scss#L64), [.cm-s-obsidian span.cm-inline-code,](https://github.com/YazanAmmar/obsidian-theme-engine/blob/d43321f7c122433735dea80e43ee56e7680334a1/src/styles/base/_editor.scss#L95-L112), [.notice,](https://github.com/YazanAmmar/obsidian-theme-engine/blob/d43321f7c122433735dea80e43ee56e7680334a1/src/styles/base/_utilities.scss#L15-L16), [.cm-search-match {](https://github.com/YazanAmmar/obsidian-theme-engine/blob/d43321f7c122433735dea80e43ee56e7680334a1/src/styles/base/_utilities.scss#L4-L13)
Please don't overwrite core styling in a plugin, add another class/data attribute so it only applies to your plugin.

---

[// Polyfill for HTMLElement.setCssProps (Obsidian 1.7+ API)](https://github.com/YazanAmmar/obsidian-theme-engine/blob/d43321f7c122433735dea80e43ee56e7680334a1/src/utils.ts#L274)
Since you'll need to update your `minAppVersion` to at least `1.8.0` (see other comments), you can remove this polyfill.

YazanAmmar commented:
> [# Obsidian Theme Engine](https://github.com/YazanAmmar/obsidian-theme-engine/blob/d43321f7c122433735dea80e43ee56e7680334a1/README.md#L1) Naming something "Obsidian xyz" is reserved for first party products we create, please change this header.
> 
> [Copyright (c) 2025 Yazan Ammar](https://github.com/YazanAmmar/obsidian-theme-engine/blob/d43321f7c122433735dea80e43ee56e7680334a1/LICENSE#L3) Please update this to the current year
> 
> ['obsidian-theme-engine',](https://github.com/YazanAmmar/obsidian-theme-engine/blob/d43321f7c122433735dea80e43ee56e7680334a1/src/core/migrations.ts#L7-L9) Please don't mess with users hotkeys or community-plugins.json on plugin load. If you have current users that need to update then you'll need to ask them to setup relevant hotkeys again. I'd recommend deleting this whole file. After your plugin is accepted and you need to make breaking changes to your plugin data schema, you can add migration logic for that (which you already have elsewhere), but please don't add migration logic for switching between plugin ids.
> 
> [const langCode = settings.language as LocaleCode;](https://github.com/YazanAmmar/obsidian-theme-engine/blob/d43321f7c122433735dea80e43ee56e7680334a1/src/i18n/strings.ts#L35), [const obsidianLang = moment.locale();](https://github.com/YazanAmmar/obsidian-theme-engine/blob/d43321f7c122433735dea80e43ee56e7680334a1/src/main.ts#L335), ["minAppVersion": "1.5.0",](https://github.com/YazanAmmar/obsidian-theme-engine/blob/d43321f7c122433735dea80e43ee56e7680334a1/manifest.json#L5), [const languageSetting = createTopSetting()](https://github.com/YazanAmmar/obsidian-theme-engine/blob/d43321f7c122433735dea80e43ee56e7680334a1/src/ui/settings/render.ts#L90-L92) This should not be a setting, use the `getLanguage()` function from the API to get the language configured in the Obsidian settings (requires `minAppVersion: "1.8.0"`)
> 
> [heading: 'Advanced settings',](https://github.com/YazanAmmar/obsidian-theme-engine/blob/d43321f7c122433735dea80e43ee56e7680334a1/src/i18n/locales/en.ts#L65), [backgroundModalSettingsTitle: 'Background settings',](https://github.com/YazanAmmar/obsidian-theme-engine/blob/d43321f7c122433735dea80e43ee56e7680334a1/src/i18n/locales/en.ts#L87) [Avoid "settings" in settings headings](https://docs.obsidian.md/Plugins/Releasing/Plugin+guidelines#Avoid+%22settings%22+in+settings+headings)
> 
> [const backgroundsPath = `${plugin.app.vault.configDir}/backgrounds`;](https://github.com/YazanAmmar/obsidian-theme-engine/blob/d43321f7c122433735dea80e43ee56e7680334a1/src/core/background-media.ts#L154) Please store this data in your plugin's config folder instead, using [`<plugin instance>.manifest.dir`](https://docs.obsidian.md/Reference/TypeScript+API/PluginManifest/dir) instead of `<vault instance>.configDir`
> 
> [callback: async () => {](https://github.com/YazanAmmar/obsidian-theme-engine/blob/d43321f7c122433735dea80e43ee56e7680334a1/src/commands/index.ts#L42), [callback: async () => {](https://github.com/YazanAmmar/obsidian-theme-engine/blob/d43321f7c122433735dea80e43ee56e7680334a1/src/commands/index.ts#L60), [callback: async () => {](https://github.com/YazanAmmar/obsidian-theme-engine/blob/d43321f7c122433735dea80e43ee56e7680334a1/src/commands/index.ts#L91) This should be a [conditional command](https://docs.obsidian.md/Plugins/User+interface/Commands#Conditional+commands)
> 
> [export function debounce(fn: (...args: unknown[]) => void, ms = 200) {](https://github.com/YazanAmmar/obsidian-theme-engine/blob/d43321f7c122433735dea80e43ee56e7680334a1/src/utils.ts#L75-L81), [debounce(func: (...args: unknown[]) => void, wait: number) {](https://github.com/YazanAmmar/obsidian-theme-engine/blob/d43321f7c122433735dea80e43ee56e7680334a1/src/ui/modals/settings/background.ts#L171-L183), [_debounce(func: (...args: unknown[]) => void, delay: number) {](https://github.com/YazanAmmar/obsidian-theme-engine/blob/d43321f7c122433735dea80e43ee56e7680334a1/src/ui/modals/editors/snippet-css.ts#L28-L34), and others Use [`debounce`](https://docs.obsidian.md/Reference/TypeScript+API/debounce) from the Obsidian API instead.
> 
> [this.colorUpdateInterval = window.setInterval(() => {](https://github.com/YazanAmmar/obsidian-theme-engine/blob/d43321f7c122433735dea80e43ee56e7680334a1/src/main.ts#L95) Events need to be registered, so they can be properly unloaded when the plugin is disabled. https://docs.obsidian.md/Plugins/Events
> 
> [.is-ltr {](https://github.com/YazanAmmar/obsidian-theme-engine/blob/d43321f7c122433735dea80e43ee56e7680334a1/src/styles/base/_direction.scss#L2), [.is-rtl {](https://github.com/YazanAmmar/obsidian-theme-engine/blob/d43321f7c122433735dea80e43ee56e7680334a1/src/styles/base/_direction.scss#L22) Please update your classes to be more specific to your plugin to avoid conflicts with other plugins.
> 
> [.tag {](https://github.com/YazanAmmar/obsidian-theme-engine/blob/d43321f7c122433735dea80e43ee56e7680334a1/src/styles/base/_editor.scss#L51), [.cm-s-obsidian span.cm-hashtag {](https://github.com/YazanAmmar/obsidian-theme-engine/blob/d43321f7c122433735dea80e43ee56e7680334a1/src/styles/base/_editor.scss#L64), [.cm-s-obsidian span.cm-inline-code,](https://github.com/YazanAmmar/obsidian-theme-engine/blob/d43321f7c122433735dea80e43ee56e7680334a1/src/styles/base/_editor.scss#L95-L112), [.notice,](https://github.com/YazanAmmar/obsidian-theme-engine/blob/d43321f7c122433735dea80e43ee56e7680334a1/src/styles/base/_utilities.scss#L15-L16), [.cm-search-match {](https://github.com/YazanAmmar/obsidian-theme-engine/blob/d43321f7c122433735dea80e43ee56e7680334a1/src/styles/base/_utilities.scss#L4-L13) Please don't overwrite core styling in a plugin, add another class/data attribute so it only applies to your plugin.
> 
> [// Polyfill for HTMLElement.setCssProps (Obsidian 1.7+ API)](https://github.com/YazanAmmar/obsidian-theme-engine/blob/d43321f7c122433735dea80e43ee56e7680334a1/src/utils.ts#L274) Since you'll need to update your `minAppVersion` to at least `1.8.0` (see other comments), you can remove this polyfill.

Hello,
Thank you for the thorough review. I have addressed all the points mentioned:

Changed the README header to remove the 'Obsidian xyz' naming.

Updated the copyright year in the LICENSE.

Completely removed the migration logic (src/core/migrations.ts) to avoid touching system config files and hotkeys.

Removed the manual language setting. The plugin now uses Obsidian's native getLanguage() API.

Removed 'Settings' from all settings headings.

Moved the backgrounds folder storage to use manifest.dir.

Converted the specified commands to conditional commands.

Replaced custom debounce implementations with the native Obsidian API.

Properly registered colorUpdateInterval.

Added plugin-specific prefixes to CSS classes to prevent bleeding, and scoped editor overrides tightly.

Removed the setCssProps polyfill and bumped minAppVersion to 1.8.7 for native getLanguage() support.

All changes have been pushed. Please let me know if anything else is needed.
